src/wasm/wasm-type.cpp

```diff
@@ -3212,10 +3224,59 @@ std::optional<TypeBuilder::Error> canonicalizeIsorecursive(
     }
   }
 
-  // TODO: proper isorecursive validation and canonicalization. For now just
-  // piggyback on the nominal system.
-  if (auto err = canonicalizeNominal(state)) {
-    return err;
+  // Check that supertypes precede their subtypes and that other child types
+  // either precede their parents or appear later in the same recursion group.
+  // `indexOfType` both maps types to their indices and keeps track of which
+  // types we have seen so far.
+  std::unordered_map<HeapType, size_t> indexOfType;
+  std::optional<RecGroup> currGroup;
+  size_t groupStart = 0;
+
+  // Validate the children of all types in a recursion group after all the types
+  // have been registered in `indexOfType`.
+  auto finishGroup = [&](size_t groupEnd) -> std::optional<TypeBuilder::Error> {
+    for (size_t index = groupStart; index < groupEnd; ++index) {
+      HeapType type = state.results[index];
+      for (HeapType child : type.getHeapTypeChildren()) {
+        // Basic children, globally canonical children, and children defined in
+        // this or previous recursion groups are allowed.
+        if (!isTemp(child) || indexOfType.count(child)) {
+          continue;
```

Comment:
How about flipping this condition? That is, "if error, error". Then we don't need the continue and the code would be slightly simpler.